// @gorhom/bottom-sheet/lib/commonjs/components/bottomSheetBackdrop/BottomSheetBackdrop.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _reactNative = require("react-native");

var _reactNativeReanimated = _interopRequireWildcard(require("react-native-reanimated"));

var _lodash = _interopRequireDefault(require("lodash.isequal"));

var _reactNativeRedash = require("react-native-redash");

var _useBottomSheet = require("../../hooks/useBottomSheet");

var _constants = require("./constants");

var _constants2 = require("../../constants");

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

const {
  interpolate: interpolateV1,
  interpolateNode: interpolateV2
} = require('react-native-reanimated');

const interpolate = interpolateV2 || interpolateV1;

const AnimatedTouchableWithoutFeedback = _reactNativeReanimated.default.createAnimatedComponent(_reactNative.TouchableWithoutFeedback);

const BottomSheetBackdropComponent = ({
  animatedIndex,
  animatedPosition,
  opacity = _constants.DEFAULT_OPACITY,
  appearsOnIndex = _constants.DEFAULT_APPEARS_ON_INDEX,
  disappearsOnIndex = _constants.DEFAULT_DISAPPEARS_ON_INDEX,
  enableTouchThrough = _constants.DEFAULT_ENABLE_TOUCH_THROUGH,
  closeOnPress = _constants.DEFAULT_CLOSE_ON_PRESS,
  style
}) => {
  //#region hooks
  const {
    close
  } = (0, _useBottomSheet.useBottomSheet)(); //#endregion
  //#region variables

  const containerRef = (0, _react.useRef)(null);
  const pointerEvents = (0, _react.useMemo)(() => enableTouchThrough ? 'none' : 'auto', [enableTouchThrough]); //#endregion
  //#region animation variables

  const isTouchable = (0, _reactNativeRedash.useValue)(closeOnPress !== undefined ? 1 : 0);
  const animatedOpacity = (0, _react.useMemo)(() => interpolate(animatedIndex, {
    inputRange: [disappearsOnIndex, appearsOnIndex],
    outputRange: [0, opacity],
    extrapolate: _reactNativeReanimated.Extrapolate.CLAMP
  }), [animatedIndex, opacity, appearsOnIndex, disappearsOnIndex]); //#endregion
  //#region callbacks

  const handleOnPress = (0, _react.useCallback)(() => {
    close();
  }, [close]); //#endregion
  //#region styles

  const buttonStyle = (0, _react.useMemo)(() => [style, {
    top: (0, _reactNativeReanimated.cond)((0, _reactNativeReanimated.eq)(animatedIndex, disappearsOnIndex), _constants2.WINDOW_HEIGHT, 0)
  }], [disappearsOnIndex, style, animatedIndex]);
  const containerStyle = (0, _react.useMemo)(() => [style, _styles.styles.container, {
    opacity: animatedOpacity
  }], [style, animatedOpacity]); //#endregion

  return closeOnPress ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(AnimatedTouchableWithoutFeedback, {
    accessible: true,
    accessibilityRole: "button",
    accessibilityLabel: "Bottom Sheet backdrop",
    accessibilityHint: "Tap to close the Bottom Sheet",
    onPress: handleOnPress,
    style: buttonStyle
  }, /*#__PURE__*/_react.default.createElement(_reactNativeReanimated.default.View, {
    ref: containerRef,
    style: containerStyle
  })), /*#__PURE__*/_react.default.createElement(_reactNativeReanimated.default.Code, null, () => (0, _reactNativeReanimated.cond)((0, _reactNativeReanimated.and)((0, _reactNativeReanimated.eq)(animatedPosition, disappearsOnIndex), isTouchable), [(0, _reactNativeReanimated.set)(isTouchable, 0), (0, _reactNativeReanimated.call)([], () => {
    // @ts-ignore
    containerRef.current.setNativeProps({
      pointerEvents: 'none'
    });
  })], (0, _reactNativeReanimated.cond)((0, _reactNativeReanimated.and)((0, _reactNativeReanimated.neq)(animatedPosition, disappearsOnIndex), (0, _reactNativeReanimated.not)(isTouchable)), [(0, _reactNativeReanimated.set)(isTouchable, 1), (0, _reactNativeReanimated.call)([], () => {
    // @ts-ignore
    containerRef.current.setNativeProps({
      pointerEvents: 'auto'
    });
  })])))) : /*#__PURE__*/_react.default.createElement(_reactNativeReanimated.default.View, {
    pointerEvents: pointerEvents,
    style: containerStyle
  });
};

const BottomSheetBackdrop = /*#__PURE__*/(0, _react.memo)(BottomSheetBackdropComponent, _lodash.default);
var _default = BottomSheetBackdrop;
exports.default = _default;
